controller/resource.go

```diff
@@ -174,17 +173,62 @@ func (t *Task) Upsert() error {
 	}
 
 	// Otherwise, create one.
-	t.Log.Info("creating service token", "id", serviceName)
 	err = t.updateServiceToken(serviceName, secretName)
 	if err != nil {
 		return fmt.Errorf("updating service token: %w", err)
 	}
-	t.Log.Info("service token created successfully", "id", serviceName)
 
 	return nil
 }
 
 func (t *Task) Delete() error {
+	if !t.isMeshTask() {
+		t.Log.Info("skipping non-mesh task", "id", *t.Task.TaskDefinitionArn)
+		return nil
+	}
+
+	tokenList, _, err := t.ConsulClient.ACL().TokenList(nil)
+	if err != nil {
+		return fmt.Errorf("listing tokens: %w", err)
+	}
+
+	var tokenFound bool
+	var serviceName string
+	for _, tokenEntry := range tokenList {
+		token, _, err := t.ConsulClient.ACL().TokenRead(tokenEntry.AccessorID, nil)
+		if err != nil {
+			return fmt.Errorf("reading token: %w", err)
+		}
+
+		serviceName, err = t.parseFamilyNameFromTaskDefinitionARN()
+		if err != nil {
+			return fmt.Errorf("parsing service name: %w", err)
+		}
+
+		if len(token.ServiceIdentities) == 1 && token.ServiceIdentities[0].ServiceName == serviceName {
+			tokenFound = true
+			t.Log.Info("deleting token", "service", serviceName)
+			_, err = t.ConsulClient.ACL().TokenDelete(token.AccessorID, nil)
+			if err != nil {
+				return fmt.Errorf("deleting token: %w", err)
+			}
+			t.Log.Info("token deleted successfully", "service", serviceName)
+
+			secretName := t.secretName(serviceName)
+			t.Log.Info("updating secret", "name", secretName)
+			_, err = t.SecretsManagerClient.UpdateSecret(&secretsmanager.UpdateSecretInput{
+				SecretId:     aws.String(secretName),
+				SecretString: aws.String(`{}`),
+			})
+			if err != nil {
+				return fmt.Errorf("updating secret: %s", err)
+			}
+			t.Log.Info("secret updated successfully", "name", secretName)
+		}
+	}
+	if !tokenFound {
+		return fmt.Errorf("could not find token for service %q", serviceName)
```

Comment:
Should this return an error? Doesn't this mean that we're 1/2 way through deleting and we should double check that the SM secret is also zero'd out?

For example, what if our consul token delete succeeds but then the SM update fails. Then we'll be in an infinite loop of trying to delete the token.

Reply:
🤔 this if statement is outside the loop. We will return an error if we looked through all tokens and haven't found one for the service. Not sure what you mean.

Reply:
I created https://github.com/hashicorp/consul-ecs/pull/31 to better illustrate. It's a bit of an edge case but I'm asking what happens if the `UpdateSecret` call fails for some reason.

Reply:
Ooh, great catch!! Thanks so much for creating the code example too, that was super helpful ❤️ 

I think you're right that this error doesn't make sense and that we should instead make behavior idempotent, in that if there's no token, we should still update existing secret. I've updated the code and added test for this use case.

Reply:
Sorry, I don't see that test case yet?

Reply:
Sorry my comment was confusing. I ended up modifying the existing test case when a token doesn't exist in consul but I realized after looking at it now that I need to update it to test the case when token doesn't exist but the secret has some value.

Reply:
Updated. I've split into different scenarios covering permutations of either token existing or secret having some value in it: 

https://github.com/hashicorp/consul-ecs/blob/ca1b6473f74f891207b8a893fe5a5bbafb11553f/controller/resource_test.go#L196-L235